### ISSUE TITLE
[Bug] a2a3 rejects row-broadcast weight slices lowered as 16x1 ColMajor Vec TLOAD from ND GM

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto#1230]
### Diagnosis

**pypto / a2a3 codegen** — the frontend lowers a valid tensor expression using `pl.slice(..., [T_TILE, 1])` + `pl.row_expand_mul` into a `16x1` `ColMajor` Vec tile load from an ND global tensor. The generated kernel passes on `a2a3sim`, but real `a2a3` incore compilation rejects the generated TLOAD:

```text
TLoad.hpp:459: error: static assertion failed due to requirement 'isSameLayout':
Fix: TLOAD(VecTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ!
```

### Description

`DeepSeekV4DecodeHcPost` needs to multiply `[T_TILE, D_CHUNK]` data blocks by per-token weights `[T_TILE, 1]`, without reading `post`/`comb` weights as scalar values. The natural tensor implementation is:

```python
post_w = pl.slice(post_2d, [T_TILE, 1], [t0, out_h])
x_rows = pl.slice(x_flat, [T_TILE, D_CHUNK], [t0, d0])
y_rows = pl.row_expand_mul(x_rows, post_w)

comb_w = pl.slice(comb_2d, [T_TILE, 1], [t0, in_h * HC_MULT + out_h])
residual_rows = pl.slice(residual_flat, [T_TILE, D_CHUNK], [t0, in_h * D + d0])
y_rows = pl.add(y_rows, pl.row_expand_mul(residual_rows, comb_w))
```

This compiles and runs correctly on `a2a3sim`, but fails during real `a2a3` kernel compilation because the weight slices lower to `16x1` `ColMajor` Vec tiles loaded from ND global tensors.

This appears related to, but distinct from, #1212. #1212 tracks silent wrong results for strided slices feeding matmul. This issue is a compile-time failure for vector row-broadcast weights.

#### Reproduction Script

Save as `repro_hc_post_colmajor_tload.py` in `pypto-lib` and run the commands below.

```python
import pypto.language as pl

B = 16
S = 1
D = 4096
HC_MULT = 4
T = B * S
T_TILE = 16
D_CHUNK = 512
D_BLOCKS = D // D_CHUNK
HC_DIM = HC_MULT * D


def build_program():
    @pl.program
    class ReproHcPostColMajorTload:
        @pl.function(type=pl.FunctionType.Opaque)
        def repro(
            self,
            x:        pl.Tensor[[B, S, D],                 pl.BF16],
            residual: pl.Tensor[[B, S, HC_MULT, D],        pl.BF16],
            post:     pl.Tensor[[B, S, HC_MULT],           pl.FP32],
            comb:     pl.Tensor[[B, S, HC_MULT, HC_MULT],  pl.FP32],
            y:        pl.Out[pl.Tensor[[B, S, HC_MULT, D], pl.BF16]],
        ):
            x_flat = pl.reshape(x, [T, D])
            residual_flat = pl.reshape(residual, [T, HC_DIM])
            post_2d = pl.reshape(post, [T, HC_MULT])
            comb_2d = pl.reshape(comb, [T, HC_MULT * HC_MULT])
            y_flat = pl.reshape(y, [T, HC_DIM])

            for out_h in pl.parallel(HC_MULT):
                with pl.at(level=pl.Level.CORE_GROUP, optimization=pl.chunked_loop_optimizer, name_hint="hc_post"):
                    for t0 in pl.parallel(0, T, T_TILE, chunk=1):
                        post_w = pl.slice(post_2d, [T_TILE, 1], [t0, out_h])
                        for db in pl.range(D_BLOCKS):
                            d0 = db * D_CHUNK
                            x_rows = pl.cast(
                                pl.slice(x_flat, [T_TILE, D_CHUNK], [t0, d0]),
                                target_type=pl.FP32,
                            )
                            y_rows = pl.row_expand_mul(x_rows, post_w)
                            for in_h in pl.range(HC_MULT):
                                comb_w = pl.slice(
                                    comb_2d,
                                    [T_TILE, 1],
                                    [t0, in_h * HC_MULT + out_h],
                                )
                                residual_rows = pl.cast(
                                    pl.slice(residual_flat, [T_TILE, D_CHUNK], [t0, in_h * D + d0]),
                                    target_type=pl.FP32,
                                )
                                y_rows = pl.add(y_rows, pl.row_expand_mul(residual_rows, comb_w))
                            y_flat = pl.assemble(
                                y_flat,
                                pl.cast(y_rows, target_type=pl.BF16),
                                [t0, out_h * D + d0],
                            )
            y = pl.reshape(y_flat, [B, S, HC_MULT, D])
            return y

    return ReproHcPostColMajorTload


def golden_fn(tensors):
    import torch

    x = tensors["x"].float()
    residual = tensors["residual"].float()
    post = tensors["post"].float()
    comb = tensors["comb"].float()

    term1 = post.unsqueeze(-1) * x.unsqueeze(-2)
    term2 = (comb.unsqueeze(-1) * residual.unsqueeze(-2)).sum(dim=2)
    tensors["y"][:] = (term1 + term2).to(torch.bfloat16)


def build_tensor_specs():
    import torch
    from golden import TensorSpec

    def init_x():
        return torch.randn(B, S, D) * 0.1

    def init_residual():
        return torch.randn(B, S, HC_MULT, D) * 0.1

    def init_post():
        p = torch.rand(B, S, HC_MULT) + 0.1
        return p / p.sum(dim=-1, keepdim=True)

    def init_comb():
        c = torch.rand(B, S, HC_MULT, HC_MULT) + 0.1
        return c / c.sum(dim=-1, keepdim=True)

    return [
        TensorSpec("x",        [B, S, D],                 torch.bfloat16, init_value=init_x),
        TensorSpec("residual", [B, S, HC_MULT, D],        torch.bfloat16, init_value=init_residual),
        TensorSpec("post",     [B, S, HC_MULT],           torch.float32,  init_value=init_post),
        TensorSpec("comb",     [B, S, HC_MULT, HC_MULT],  torch.float32,  init_value=init_comb),
        TensorSpec("y",        [B, S, HC_MULT, D],        torch.bfloat16, is_output=True),
    ]


if __name__ == "__main__":
    import argparse
    from golden import RunConfig, run

    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", type=str, default="a2a3",
                        choices=["a2a3", "a2a3sim", "a5", "a5sim"])
    parser.add_argument("-d", "--device", type=int, default=0)
    args = parser.parse_args()

    result = run(
        program=build_program(),
        tensor_specs=build_tensor_specs(),
        golden_fn=golden_fn,
        config=RunConfig(
            rtol=3e-3,
            atol=3e-3,
            compile=dict(dump_passes=True),
            runtime=dict(platform=args.platform, device_id=args.device),
        ),
    )
    if not result.passed:
        if result.error:
            print(result.error)
        raise SystemExit(1)
```

#### Commands and Results

Simulator passes:

```bash
python repro_hc_post_colmajor_tload.py -p a2a3sim
```

Observed result:

```text
[RUN]   'y' PASS  shape=(16, 1, 4, 4096) dtype=torch.bfloat16
[RUN] PASS
```

Real a2a3 fails during incore kernel compilation:

```bash
task-submit --device auto --max-time 120 --timeout 180 \
  --env PYTHONPATH --env ASCEND_HOME_PATH --env PTOAS_ROOT --env PTO_ISA_ROOT \
  --run "python repro_hc_post_colmajor_tload.py -p a2a3 -d {}"
```

Observed error:

```text
[RUN] runtime ...
[Incore] Compilation failed:
...
TLoad.hpp:459:9: error: static assertion failed due to requirement 'isSameLayout':
Fix: TLOAD(VecTile, GlobalTensor) only support ND2ND/DN2DN/NZ2NZ!

note: in instantiation of function template specialization
TLOAD_TILE_IMPL<
  Tile<Vec, float, 16, 1, BLayout::ColMajor, ...>,
  GlobalTensor<float, Shape<1,1,1,16,1>, Stride<64,64,64,4,1>, Layout::ND>
>
...
TLOAD(v29, v34);  // post_w

TLOAD_TILE_IMPL<
  Tile<Vec, float, 16, 1, BLayout::ColMajor, ...>,
  GlobalTensor<float, Shape<1,1,1,16,1>, Stride<256,256,256,16,1>, Layout::ND>
>
...
TLOAD(v55, v60);  // comb_w
```

Generated PTO excerpt:

```mlir
%post_w__tile = pto.alloc_tile ... :
  !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=1, blayout=col_major, ...>
%post_2d__ssa_v0_pview = pto.partition_view %post_2d__ssa_v0_view,
  offsets = [%6, %arg6], sizes = [%c16_index, %c1_index]
pto.tload ins(%post_2d__ssa_v0_pview) outs(%post_w__tile)

%comb_w__tile = pto.alloc_tile ... :
  !pto.tile_buf<loc=vec, dtype=f32, rows=16, cols=1, blayout=col_major, ...>
%comb_2d__ssa_v0_pview = pto.partition_view %comb_2d__ssa_v0_view,
  offsets = [%9, %11], sizes = [%c16_index, %c1_index]
pto.tload ins(%comb_2d__ssa_v0_pview) outs(%comb_w__tile)
```

Expected behavior:

1. `pl.slice(..., [T_TILE, 1])` used as a row-broadcast weight should compile and run correctly on real `a2a3`, as it does on `a2a3sim`.
2. If a2a3 cannot load an ND GM view directly into a `ColMajor` Vec tile, pypto/codegen should lower through a supported layout, insert an appropriate transform/repack, or emit an earlier actionable diagnostic.

### Environment

| Component | Version |
|---|---|
| pypto-lib | `2428832` |
| pypto | `8a72fc35` (branch: `main`) |
| simpler | `08f6f769` |
| ptoas | `0.33` |
| CANN | `8.5.0.alpha001` |

### Host Platform

Linux (aarch64)

### Additional Context

The scalar version using `pl.read(post_flat, ...)` and `pl.read(comb_flat, ...)` compiles and passes, but it forces scalar weight paths. The tensor version is needed to keep the HC post mix vectorized over `T_TILE` rows.

Related: #1212

**Comment by lyfne123** [comment]
## Working version using existing supported patterns

经过对硬件/PTOAS 限制的进一步排查,这个 issue 描述的 `pl.slice(post_2d, [T_TILE, 1], [t0, h])` 模式在 a2a3 上**没有有效的 tile 表示**:

- `ColMajor [16, 1]` + ND 源 → `isSameLayout` 失败(本 issue 报错)
- `RowMajor [16, 1]` + ND 源 → `pto.alloc_tile` 行字节对齐失败(`cols * sizeof(FP32) = 4B < 32B`)
- 把 ND 列投影伪装成 DN 视图 → 列在物理内存上的 stride 是 `HC=4`,而 PTOAS DN canonical stride 必须是 `1`,不能仅靠 metadata retag 完成

也就是 **`[N, 1]` FP32 tile 只能从"PTOAS 已经能 DN-tag 的源"加载**(即源 tensor 自身 shape 末尾是 1)。这是 PTOAS 当前实现的硬约束。

---

下面这个版本**只用 repo 里已有受支持的 op**(无需任何 PyPTO 改动),在保留向量化语义、不退到 scalar 路径的前提下,可以在真机 a2a3 跑通。

### 核心改写思路

| | 原写法(挂) | 新写法(通) |
|---|---|---|
| **数据物理布局** | `post: [B, S, HC]` 内存原样 | 模型加载时一次性 transpose 成 `post_t: [HC, T]` |
| **per-h 权重在内存里的形态** | 第 h 列(stride `HC`,非连续) | 第 h 行(stride 1,连续 `T × sizeof(dtype)` 字节) |
| **kernel 内取权重** | `pl.slice(post_2d, [T_TILE, 1], [t0, h])` ← 触发 bug | `pl.slice(post_t, [1, T_TILE], [h, t0])` → `pl.reshape(..., [T_TILE, 1])` |
| **`row_expand_mul` 输入** | 期望 `[T_TILE, 1]` ColMajor | **完全相同** |

### 完整 kernel

```python
import pypto.language as pl

B, S, D = 16, 1, 4096
HC_MULT = 4
T = B * S          # 16
T_TILE = 16
D_CHUNK = 512
D_BLOCKS = D // D_CHUNK
HC_DIM = HC_MULT * D


@pl.program
class HcPostFixed:
    @pl.function(type=pl.FunctionType.Opaque)
    def repro(
        self,
        x:        pl.Tensor[[B, S, D],                 pl.BF16],
        residual: pl.Tensor[[B, S, HC_MULT, D],        pl.BF16],
        # 一次性预转置:post_t[h, t] = original_post[t // S, t % S, h]
        # 每行是某个 h 对应 T 个连续 FP32 浮点
        post_t:   pl.Tensor[[HC_MULT, T],              pl.FP32],
        # 同理:comb_t[h_in * HC_MULT + h_out, t] = original_comb[t // S, t % S, h_in, h_out]
        comb_t:   pl.Tensor[[HC_MULT * HC_MULT, T],    pl.FP32],
        y:        pl.Out[pl.Tensor[[B, S, HC_MULT, D], pl.BF16]],
    ):
        x_flat        = pl.reshape(x,            [T, D])
        residual_flat = pl.reshape(residual,     [T, HC_DIM])
        y_flat        = pl.reshape(y,            [T, HC_DIM])

        for out_h in pl.parallel(HC_MULT):
            with pl.at(level=pl.Level.CORE_GROUP,
                       optimization=pl.chunked_loop_optimizer,
                       name_hint="hc_post"):
                for t0 in pl.parallel(0, T, T_TILE, chunk=1):
                    # 取 post_t 的第 out_h 行作为 [1, T_TILE] RowMajor tile
                    # (row byte = 16 × 4 = 64B,满足 PTOAS 32B 对齐;ND2ND TLOAD)
                    post_row = pl.slice(post_t, [1, T_TILE], [out_h, t0])
                    # 再 reshape 成 [T_TILE, 1] ColMajor —— 纯 metadata retag,
                    # 物理 buffer 不变,tile.reshape 对 cols=1 自动 infer col_major
                    post_w   = pl.reshape(post_row, [T_TILE, 1])

                    for db in pl.range(D_BLOCKS):
                        d0 = db * D_CHUNK
                        x_rows = pl.cast(
                            pl.slice(x_flat, [T_TILE, D_CHUNK], [t0, d0]),
                            target_type=pl.FP32,
                        )
                        y_rows = pl.row_expand_mul(x_rows, post_w)

                        for in_h in pl.range(HC_MULT):
                            comb_row = pl.slice(
                                comb_t, [1, T_TILE], [in_h * HC_MULT + out_h, t0]
                            )
                            comb_w = pl.reshape(comb_row, [T_TILE, 1])

                            residual_rows = pl.cast(
                                pl.slice(residual_flat,
                                         [T_TILE, D_CHUNK],
                                         [t0, in_h * D + d0]),
                                target_type=pl.FP32,
                            )
                            y_rows = pl.add(
                                y_rows,
                                pl.row_expand_mul(residual_rows, comb_w),
                            )

                        y_flat = pl.assemble(
                            y_flat,
                            pl.cast(y_rows, target_type=pl.BF16),
                            [t0, out_h * D + d0],
                        )

        y = pl.reshape(y_flat, [B, S, HC_MULT, D])
        return y
```

### 一次性数据预处理(在模型加载时)

```python
# 在 PyTorch / numpy 层做一次,不在 decode 热路径上
post_t = post.permute(2, 0, 1).contiguous().reshape(HC_MULT, T)
comb_t = comb.permute(2, 3, 0, 1).contiguous().reshape(HC_MULT * HC_MULT, T)
```

`B=16, S=1, HC=4` 这个规模下,`post` 总共 256 字节、`comb` 1024 字节,转置成本可以忽略。

### IR / 后端层面为什么这次通了

我在本地把简化版跑通过 codegen,关键 PTO IR 片段如下:

| 步骤 | IR 形状 | tile_buf | TLOAD | 对齐 |
|---|---|---|---|---|
| `pl.slice(post_t, [1, T_TILE], [out_h, t0])` | `[1, 16]` | RowMajor | **ND2ND** ✓ | row byte = 64B ✓ |
| `pl.reshape(post_row, [T_TILE, 1])` | `[16, 1]` | **ColMajor** (auto infer) | 无 TLOAD,**复用同一 buffer addr**(纯 metadata retag) | col byte = 64B ✓ |
| `pl.row_expand_mul(x_rows, post_w)` | data `[16, 512]` RowMajor + weight `[16, 1]` ColMajor | — | 同 `test_pto_codegen_paged_attn.py` `oi_scaled = pl.tile.row_expand_mul(oi_tile, alpha_tile)` | — |

`row_expand_mul` 接受 ColMajor row_vec + RowMajor data tile 这个组合在 `examples/models/04_paged_attention.py:205-208` 和 `examples/models/09_paged_attention_spmd.py:395-398` 都有现成用例,是 production-tested 的模式:

```python
alpha_dn = pl.reshape(alpha, [16, 1])   # Reshape [1,16] -> [16,1] DN for row_expand_mul
oi_scaled = pl.row_expand_mul(oi_tile, alpha_dn)
```

### 限制与后续

- 这个改写要求 `T_TILE × sizeof(dtype) ≥ 32`。当前 `T_TILE=16` + FP32(64B)、BF16(32B)都满足。如果未来 `T_TILE < 8`(FP32)或 `< 16`(BF16)就会再次撞 PTOAS 对齐限制,届时需要进一步把 `T` 维 padding/tile 切大。
- PR #1438 已经把 IR 层"ColMajor 收紧到只对 DN-equivalent 源生效"修了(避免**不**符合 DN-equivalent 的源仍被错误标为 ColMajor 的隐性 bug),但没有也无法在 IR 层让原 `pl.slice(post_2d, [T_TILE, 1], [t0, h])` 写法跑通 —— 因为 a2a3 PTOAS 不支持那种 tile 表示。
- 长期看,**[Option B] 在 IR pass 层把这类 pattern 自动改写到上面这种 transpose-once 形式**是更好的用户体验,需要单开一个 RFC 评估。
